In [1]:
from IPython.display import display, HTML
display(HTML("<style>.container { width:98% !important; }</style>"))

In [2]:
#Import Libraries
import numpy as np
import pandas as pd

In [3]:
dataset = pd.read_csv('D600 Task 2 Dataset 1 Housing Information.csv')
dataset = dataset[['Price', 'SquareFootage', 'CrimeRate', 'RenovationQuality', 'Fireplace', 'IsLuxury']]
dataset

Price  SquareFootage  CrimeRate  RenovationQuality Fireplace  \
0     255614.8992         566.62      20.56               4.93       Yes   
1     155586.0947        1472.34      15.62               4.08        No   
2     131050.8324         550.00      12.47               4.26       Yes   
3     151361.7125         941.81      22.22               4.45       Yes   
4     113167.6128         550.00       8.28               3.36        No   
...           ...            ...        ...                ...       ...   
6995  307821.1758         550.00      11.89               4.76        No   
6996  421368.8869        1726.12      34.06               5.31        No   
6997  473382.5348        1026.36       0.10               4.98        No   
6998  343397.9756        2218.22      11.18               4.68        No   
6999  438060.8193        1553.57      18.80               8.80        No   

      IsLuxury  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
6995         1  
6996         1  
6997         1  
6998         0  
6999         1  

[7000 rows x 6 columns]

In [4]:
#Preprocessing for the model
df = dataset.copy()
df = pd.get_dummies(df, drop_first=True)
# df.head()

#Split dataset into X features and y label
X = df.drop(columns=['IsLuxury'])
y = df['IsLuxury']

import statsmodels.api as sm 
X = sm.add_constant(X) # add one constant column for the intercept
X.head()

const        Price  SquareFootage  CrimeRate  RenovationQuality  \
0    1.0  255614.8992         566.62      20.56               4.93   
1    1.0  155586.0947        1472.34      15.62               4.08   
2    1.0  131050.8324         550.00      12.47               4.26   
3    1.0  151361.7125         941.81      22.22               4.45   
4    1.0  113167.6128         550.00       8.28               3.36   

   Fireplace_Yes  
0              1  
1              0  
2              1  
3              1  
4              0

In [5]:
#-----------------Model 1

#Split dataset into train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.to_csv('Task2_X_train.csv', index=False)
X_test.to_csv('Task2_X_test.csv', index=False)
y_train.to_csv('Task2_y_train.csv', index=False)
y_test.to_csv('Task2_y_test.csv', index=False)

#D2
import statsmodels.api as sm
model = sm.Logit(y_train, X_train).fit()  

print(model.summary())
print(model.params.round(10))

print("AIC:", model.aic)
print("BIC:", model.bic)

Optimization terminated successfully.
         Current function value: 0.430256
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:               IsLuxury   No. Observations:                 5600
Model:                          Logit   Df Residuals:                     5594
Method:                           MLE   Df Model:                            5
Date:                Sat, 01 Feb 2025   Pseudo R-squ.:                  0.3792
Time:                        12:33:09   Log-Likelihood:                -2409.4
converged:                       True   LL-Null:                       -3881.2
Covariance Type:            nonrobust   LLR p-value:                     0.000
                        coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -4.6268      0.162    -28.578      0.000      -4.944      -4.310
Price   

Note:
- p_value for RenovationQuality is not statistically significant. Lets remove that in the next model. 
- Write down the AIC and BIC (the lower these numbers go the better the  model becomes)

In [7]:
#Preprocessing for the model
df = dataset.copy()
df = pd.get_dummies(df, drop_first=True)
df.drop(columns='RenovationQuality', inplace=True)  #eliminating RenovationQuality 
#df.head()

#Split dataset into X features and y label
X = df.drop(columns=['IsLuxury'])
y = df['IsLuxury']

import statsmodels.api as sm 
X = sm.add_constant(X) # add one constant column for the intercept
print(X.head())


#-----------------Model 2

#Split dataset into train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#D2
import statsmodels.api as sm
model = sm.Logit(y_train, X_train).fit()  

print(model.summary())
print(model.params.round(10))

print("\nAIC:", model.aic)
print("BIC:", model.bic)

   const        Price  SquareFootage  CrimeRate  Fireplace_Yes
0    1.0  255614.8992         566.62      20.56              1
1    1.0  155586.0947        1472.34      15.62              0
2    1.0  131050.8324         550.00      12.47              1
3    1.0  151361.7125         941.81      22.22              1
4    1.0  113167.6128         550.00       8.28              0
Optimization terminated successfully.
         Current function value: 0.430257
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:               IsLuxury   No. Observations:                 5600
Model:                          Logit   Df Residuals:                     5595
Method:                           MLE   Df Model:                            4
Date:                Sat, 01 Feb 2025   Pseudo R-squ.:                  0.3792
Time:                        12:35:11   Log-Likelihood:                -2409.4
converged:                       True   LL-Nul

Note:
- p_value for Fireplace_Yes is still high, remove it
- note AIC and BIC are slowly dropping

In [8]:
#Preprocessing for the model
df = dataset.copy()
df = pd.get_dummies(df, drop_first=True)
df.drop(columns=['RenovationQuality', 'Fireplace_Yes'], inplace=True)  #adding to elimination Fireplace_Yes
#df.head()

#Split dataset into X features and y label
X = df.drop(columns=['IsLuxury'])
y = df['IsLuxury']

import statsmodels.api as sm 
X = sm.add_constant(X) # add one constant column for the intercept
print(X.head())


#-----------------Model 3

#Split dataset into train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#D2
import statsmodels.api as sm
model = sm.Logit(y_train, X_train).fit()  

print(model.summary())
print(model.params.round(10))

print("\nAIC:", model.aic)
print("BIC:", model.bic)

   const        Price  SquareFootage  CrimeRate
0    1.0  255614.8992         566.62      20.56
1    1.0  155586.0947        1472.34      15.62
2    1.0  131050.8324         550.00      12.47
3    1.0  151361.7125         941.81      22.22
4    1.0  113167.6128         550.00       8.28
Optimization terminated successfully.
         Current function value: 0.430276
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:               IsLuxury   No. Observations:                 5600
Model:                          Logit   Df Residuals:                     5596
Method:                           MLE   Df Model:                            3
Date:                Sat, 01 Feb 2025   Pseudo R-squ.:                  0.3792
Time:                        12:35:52   Log-Likelihood:                -2409.5
converged:                       True   LL-Null:                       -3881.2
Covariance Type:            nonrobust   LLR p-value:     

Note:
- p_value for CrimeRate is still high, remove it
- note AIC and BIC are slowly dropping

In [9]:
#Preprocessing for the model
df = dataset.copy()
df = pd.get_dummies(df, drop_first=True)
df.drop(columns=['RenovationQuality', 'Fireplace_Yes', 'CrimeRate'], inplace=True)  #adding to elimination CrimeRate
#df.head()

#Split dataset into X features and y label
X = df.drop(columns=['IsLuxury'])
y = df['IsLuxury']

import statsmodels.api as sm 
X = sm.add_constant(X) # add one constant column for the intercept
print(X.head())


#-----------------Model 4

#Split dataset into train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#D2
import statsmodels.api as sm
model = sm.Logit(y_train, X_train).fit()  

print(model.summary())
print(model.params.round(10))

print("\nAIC:", model.aic)
print("BIC:", model.bic)

   const        Price  SquareFootage
0    1.0  255614.8992         566.62
1    1.0  155586.0947        1472.34
2    1.0  131050.8324         550.00
3    1.0  151361.7125         941.81
4    1.0  113167.6128         550.00
Optimization terminated successfully.
         Current function value: 0.430313
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:               IsLuxury   No. Observations:                 5600
Model:                          Logit   Df Residuals:                     5597
Method:                           MLE   Df Model:                            2
Date:                Sat, 01 Feb 2025   Pseudo R-squ.:                  0.3791
Time:                        12:36:17   Log-Likelihood:                -2409.8
converged:                       True   LL-Null:                       -3881.2
Covariance Type:            nonrobust   LLR p-value:                     0.000
                    coef    std err         

Note:
- p_value for SquareFootage is still high, remove it
- note AIC and BIC are slowly dropping

In [10]:
#Preprocessing for the model
df = dataset.copy()
df = pd.get_dummies(df, drop_first=True)
df.drop(columns=['RenovationQuality', 'Fireplace_Yes', 'CrimeRate', 'SquareFootage'], inplace=True)  #adding to elimination SquareFootage
#df.head()

#Split dataset into X features and y label
X = df.drop(columns=['IsLuxury'])
y = df['IsLuxury']

import statsmodels.api as sm 
X = sm.add_constant(X) # add one constant column for the intercept
print(X.head())


#-----------------Model 5

#Split dataset into train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#D2
import statsmodels.api as sm
model = sm.Logit(y_train, X_train).fit()  

print(model.summary())
print(model.params.round(10))

print("\nAIC:", model.aic)
print("BIC:", model.bic)

   const        Price
0    1.0  255614.8992
1    1.0  155586.0947
2    1.0  131050.8324
3    1.0  151361.7125
4    1.0  113167.6128
Optimization terminated successfully.
         Current function value: 0.430355
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:               IsLuxury   No. Observations:                 5600
Model:                          Logit   Df Residuals:                     5598
Method:                           MLE   Df Model:                            1
Date:                Sat, 01 Feb 2025   Pseudo R-squ.:                  0.3791
Time:                        12:36:36   Log-Likelihood:                -2410.0
converged:                       True   LL-Null:                       -3881.2
Covariance Type:            nonrobust   LLR p-value:                     0.000
                 coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------

Note:
- now we are left with one feature and it seems to be statistically significant
- note AIC and BIC still dropping, that is good

In [12]:
# Make predictions on test set
y_pred = model.predict(X_test).reset_index(drop=True)
compare = pd.DataFrame({'y_test': y_test.reset_index(drop=True), 'y_pred': y_pred})
compare.loc[[607, 1011, 1274, 429, 186, 1131, 516, 648, 32, 27]]

y_test    y_pred
607        0  0.431257
1011       0  0.155156
1274       0  0.065014
429        1  0.975226
186        1  0.971403
1131       1  0.966265
516        1  0.897049
648        0  0.141026
32         1  0.952760
27         1  0.784594

In [13]:
# Same as cell above just convert the probabilities into binary 1 and 0
y_pred = model.predict(X_test).reset_index(drop=True)
y_pred = np.round(y_pred).astype(int).reset_index(drop=True)
compare = pd.DataFrame({'y_test': y_test.reset_index(drop=True), 'y_pred': y_pred})
compare.loc[[607, 1011, 1274, 429, 186, 1131, 516, 648, 32, 27]]

y_test  y_pred
607        0       0
1011       0       0
1274       0       0
429        1       1
186        1       1
1131       1       1
516        1       1
648        0       0
32         1       1
27         1       1

In [14]:
# Confusion matrix for test set & accuracy 

from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
confusion_matrix = pd.DataFrame(confusion_matrix, index=["NotLuxury", "IsLuxury"], columns=["PredNotLuxury", "PredIsLuxury"])
print(confusion_matrix)

# accuracy of the test set
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

           PredNotLuxury  PredIsLuxury
NotLuxury            645            62
IsLuxury              90           603


0.8914285714285715

In [15]:
# accuracy & confusion for training set

y_pred_train = model.predict(X_train).reset_index(drop=True)
y_pred_train = np.round(y_pred_train).astype(int).reset_index(drop=True)
compare = pd.DataFrame({'y_train': y_train.reset_index(drop=True), 'y_pred_train': y_pred_train})

from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_train, y_pred_train)
confusion_matrix = pd.DataFrame(confusion_matrix, index=["NotLuxury", "IsLuxury"], columns=["PredNotLuxury", "PredIsLuxury"])
print(confusion_matrix)

# accuracy of train set
from sklearn.metrics import accuracy_score
accuracy_score(y_train, y_pred_train)

           PredNotLuxury  PredIsLuxury
NotLuxury           2506           259
IsLuxury             395          2440


0.8832142857142857

Note: 
- both training and test set came at 88% and 89% respectively, model is doing very well on predicting whether the House is luxury or not
- Predictive formula: log[P(NewHouseIsLuxury) / (1-P(NewHouseIsLuxury))] = -4.552487 + (0.000016 X NewHousePrice)